# Búsqueda binaria
Este fue el primer algoritmo que aprendí que rompía mi forma de pensar. La búsqueda binaria consiste en encontrar un número dentro de una lista ordenada, dividiendo la lista en dos cada vez que realiza la búsqueda.


Si busco un número `x = 10` en una lista `l=[3,10,2,5,7,8]`. Lo primero que tengo que hacer es ordenar la lista.

In [10]:
l=[3,10,2,5,7,8]
l_ordenada = sorted(l)
l_ordenada

[2, 3, 5, 7, 8, 10]

Ahora determino un centro para empezar mi búsqueda.

In [11]:
centro = len(l)//2
num_central = l_ordenada[centro]
num_central

7

Si mi centro es igual a `x` entonces el algoritmo se detiene. En este caso, mi centro es menor al número que estoy buscando, entonces descarto todos los números que están debajo: `[2,3,5]`

In [12]:
l_recortada = l_ordenada[centro:]
l_recortada

[7, 8, 10]

Obtengo un nuevo centro y repito el proceso.

In [13]:
centro = len(l_recortada)//2
num_central = l_recortada[centro]
num_central

8

In [14]:
l_recortada = l_recortada[centro:]
l_recortada

[8, 10]

In [15]:
centro = len(l_recortada)//2
num_central = l_recortada[centro]
num_central

10

Finalmente, mi `num_central` es igual a `x`, la búsqueda se detiene y he resuelto el problema en 3 pasos. En un algoritmo poco eficiente, usando un `for`, me hubiese tomado 6 pasos. Esto no parece la gran cosa pero se nota cuando hablamso de grandes cantidades de números.

### Función de búsqueda binaria

In [6]:
#mi función de búsqueda binaria retornará el índice donde se encuentra el número
def binary_search(x,l):
    a=0
    b=len(l)-1
    while(a<=b):
        centro = (a+b)//2
        if(l[centro]==x):
            return centro
        else:
            if(l[centro]<x):
                a = centro+1
            else:
                b = centro-1
    return -1

In [9]:
l=[3,10,2,5,7,8]
l_ordenada = sorted(l)
indice = binary_search(10,l_ordenada)
print(indice,l_ordenada[indice])

5 10


In [14]:
#Si no hay x, se imrpime -1
print(binary_search(4,l_ordenada))

-1


### Comparando con For

Ahora voy a comparar búsqueda binaria con `for` en términos de cuántas veces preguntamos por `x`.

In [22]:
l = range(1000)
x = 800
preguntas = 0
for i in range(len(l)):
    preguntas+=1
    if(x==i):
        print("Encontrado en el índice: ",i)
        break #el peor de los casos recorre toda la lista
        

print("Preguntas: ",preguntas)

Encontrado en el índice:  800
Preguntas:  801


In [24]:
l = range(1000)
x = 100000
preguntas = 0
for i in range(len(l)):
    preguntas+=1
    if(x==i):
        print("Encontrado en el índice: ",i)
        break #el peor de los casos recorre toda la lista
        

print("Preguntas: ",preguntas)

Preguntas:  1000


In [36]:
#mi función de búsqueda binaria retornará el índice donde se encuentra el número
def binary_search(x,l):
    contador=0
    a=0
    b=len(l)-1
    while(a<=b):
        contador+=1
        centro = (a+b)//2
        #print(centro)
        if(l[centro]==x):
            print("Preguntas: ",contador)
            return centro
        else:
            if(l[centro]<x):
                a = centro+1
            else:
                b = centro-1
    print("Preguntas: ",contador)
    return -1

In [37]:
l = range(1000)
x = 800
i = binary_search(x,l)
        
print("Encontrado en el índice: ",i)

Preguntas:  10
Encontrado en el índice:  800


In [39]:
l = range(1000)
x = 100000
i = binary_search(x,l)
print("Encontrado en el índice: ",i)

Preguntas:  10
Encontrado en el índice:  -1


En el for 801 y 1000 preguntas, en búsqueda binaria sólo 10 en ambos casos.